packages/edit-site/src/components/page-patterns/grid.js

```diff
@@ -1,26 +1,117 @@
 /**
  * WordPress dependencies
  */
-import { __experimentalText as Text } from '@wordpress/components';
-import { useRef } from '@wordpress/element';
-import { __, sprintf } from '@wordpress/i18n';
+import {
+	__experimentalHStack as HStack,
+	__experimentalText as Text,
+	Button,
+} from '@wordpress/components';
+import { useRef, useState, useMemo } from '@wordpress/element';
+import { __, _x, _n, sprintf } from '@wordpress/i18n';
+import { useAsyncList } from '@wordpress/compose';
 
 /**
  * Internal dependencies
  */
 import GridItem from './grid-item';
 
-const PAGE_SIZE = 100;
+const PAGE_SIZE = 20;
+
+function Pagination( { currentPage, numPages, changePage, totalItems } ) {
+	return (
+		<HStack
+			expanded={ false }
+			spacing={ 3 }
+			className="edit-site-patterns__grid-pagination"
+		>
+			<Text variant="muted">
+				{
+					// translators: %s: Total number of patterns.
+					sprintf(
+						// translators: %s: Total number of patterns.
+						_n( '%s item', '%s items', totalItems ),
+						totalItems
+					)
+				}
+			</Text>
+			<HStack expanded={ false } spacing={ 1 }>
+				<Button
+					variant="tertiary"
+					onClick={ () => changePage( 1 ) }
+					disabled={ currentPage === 1 }
+					aria-label={ __( 'First page' ) }
+				>
+					«
+				</Button>
+				<Button
+					variant="tertiary"
+					onClick={ () => changePage( currentPage - 1 ) }
+					disabled={ currentPage === 1 }
+					aria-label={ __( 'Previous page' ) }
+				>
+					‹
+				</Button>
+			</HStack>
+			<Text variant="muted">
+				{ sprintf(
+					// translators: %1$s: Current page number, %2$s: Total number of pages.
+					_x( '%1$s of %2$s', 'paging' ),
+					currentPage,
+					numPages
+				) }
+			</Text>
+			<HStack expanded={ false } spacing={ 1 }>
+				<Button
+					variant="tertiary"
+					onClick={ () => changePage( currentPage + 1 ) }
+					disabled={ currentPage === numPages }
+					aria-label={ __( 'Next page' ) }
+				>
+					›
+				</Button>
+				<Button
+					variant="tertiary"
+					onClick={ () => changePage( numPages ) }
+					disabled={ currentPage === numPages }
+					aria-label={ __( 'Last page' ) }
+				>
+					»
+				</Button>
+			</HStack>
+		</HStack>
+	);
+}
 
 export default function Grid( { categoryId, items, ...props } ) {
+	const [ currentPage, setCurrentPage ] = useState( 1 );
 	const gridRef = useRef();
+	const totalItems = items.length;
+	const pageIndex = currentPage - 1;
 
-	if ( ! items?.length ) {
+	const list = useMemo(
+		() =>
+			items.slice(
+				pageIndex * PAGE_SIZE,
+				pageIndex * PAGE_SIZE + PAGE_SIZE
+			),
+		[ pageIndex, items ]
+	);
+
+	const asyncList = useAsyncList( list, { step: 10 } );
+
+	if ( ! list?.length ) {
 		return null;
 	}
 
-	const list = items.slice( 0, PAGE_SIZE );
-	const restLength = items.length - PAGE_SIZE;
+	const numPages = Math.ceil( items.length / PAGE_SIZE );
+	const changePage = ( page ) => {
+		const scrollContainer =
+			document.getElementsByClassName( 'edit-site-patterns' );
```

Comment:
Is there any functional difference to the following?

```js
		window.scrollTo( 0, 0 );
		setCurrentPage( page );
```

Just means that we won't have to worry about DOM/class name changes down the line

Reply:
Because we are scrolling an inner div window is already (0,0) so nothing happens if you do `window.scrollTo( 0, 0 );`, the actual div that currently has overflow scrolled has to be used - at least in my testing anyway.

Reply:
Fair enough, didn't see the overflow. Thanks for the explainer!

Reply:
I did look at getting a `ref` from the parent component, but you can't pass refs down to functional components, and `forwardRefs` seem to be designed just for passing DOM refs up the chain rather than down. But let me know if you can see another option for getting a ref to that element as I agree something other than the class reference would be more robust.

Reply:
I guess what threw me was `getElementsByClassName( 'edit-site-patterns' )`... are we expecting more than one element? What if there is one and it's in the first index and doesn't appear where we expect in the DOM?

Reply:
> but you can't pass refs down to functional components

Oh! You can't do something like this?

```jsx
				<Grid
					categoryId={ categoryId }
					items={ patterns }
					aria-labelledby={ titleId }
					aria-describedby={ descriptionId }
					parentRef={ patternsListRef }
				/>
```

Didn't realize....

Reply:
Oh, forget it. Just tried `Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?`

😄